pkg/odo/cli/component/create.go

```diff
@@ -38,280 +41,276 @@ var (
-					Name:       componentName,
-					SourcePath: componentGit,
-					SourceRef:  componentGitRef,
-					SourceType: occlient.GIT,
-
-					ImageName:       componentImageName,
-					EnvVars:         componentEnvVars,
-					Ports:           componentPorts,
-					Resources:       resourceQuantity,
-					ApplicationName: applicationName,
-				},
-			)
-			odoutil.LogErrorAndExit(err, "")
-
-			// Git is the only one using BuildConfig since we need to retrieve the git
-			err = component.Build(client, componentName, applicationName, true, stdout)
-			odoutil.LogErrorAndExit(err, "")
-
-		} else if len(componentLocal) != 0 {
-			fileInfo, err := os.Stat(componentPath)
-			odoutil.LogErrorAndExit(err, "")
-			if !fileInfo.IsDir() {
-				log.Errorf("Please provide a path to the directory")
-				os.Exit(1)
-			}
+// Validate validates the create parameters
+func (co *CreateOptions) Validate() (err error) {
+	componentCnt := 0
 
-			// Create
-			err = component.CreateFromPath(
-				client,
-				occlient.CreateArgs{
-					Name:            componentName,
-					SourcePath:      componentPath,
-					SourceType:      occlient.LOCAL,
-					ImageName:       componentImageName,
-					EnvVars:         componentEnvVars,
-					Ports:           componentPorts,
-					Resources:       resourceQuantity,
-					ApplicationName: applicationName,
-				},
-			)
-			odoutil.LogErrorAndExit(err, "")
-
-		} else if len(componentBinary) != 0 {
-			// Deploy the component with a binary
-
-			// Create
-			err = component.CreateFromPath(
-				client,
-				occlient.CreateArgs{
-					Name:            componentName,
-					SourcePath:      componentPath,
-					SourceType:      occlient.BINARY,
-					ImageName:       componentImageName,
-					EnvVars:         componentEnvVars,
-					Ports:           componentPorts,
-					Resources:       resourceQuantity,
-					ApplicationName: applicationName,
-				},
-			)
-			odoutil.LogErrorAndExit(err, "")
+	if len(componentBinary) != 0 {
+		componentCnt++
+	}
+	if len(componentGit) != 0 {
+		componentCnt++
+	}
+	if len(componentLocal) != 0 {
+		componentCnt++
+	}
+	if componentCnt > 1 {
+		return fmt.Errorf("The source can be either --binary or --local or --git")
+	}
 
-		} else {
-			// If the user does not provide anything (local, git or binary), use the current absolute path and deploy it
-			dir, err := util.GetAbsPath("./")
-			odoutil.LogErrorAndExit(err, "")
-
-			// Create
-			err = component.CreateFromPath(
-				client,
-				occlient.CreateArgs{
-					Name:            componentName,
-					SourcePath:      dir,
-					SourceType:      occlient.LOCAL,
-					ImageName:       componentImageName,
-					EnvVars:         componentEnvVars,
-					Ports:           componentPorts,
-					Resources:       resourceQuantity,
-					ApplicationName: applicationName,
-				},
-			)
-			odoutil.LogErrorAndExit(err, "")
-		}
+	if len(componentGit) == 0 && len(componentGitRef) != 0 {
+		return fmt.Errorf("The --ref flag is only valid for --git flag")
```

Comment:
I slightly disagree here bcoz I think if the user has cloned a git repo and has multiple branches, he might as well expect that an equivalent of a local git checkout is done automatically and the code that is pushed is from that branch... So, it might be misleading..

> That would change the externally visible behavior, though, and might be better suited for a different PR.

I agree we can revisit it in another PR